## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [58]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [59]:
# View the column names in the data
df = pd.read_csv('Resources/client_dataset.csv')
df.head()

df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [60]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [12]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [63]:
# What three item categories had the most entries?
df["category"].value_counts().head(3)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [67]:
# For the category with the most entries, which subcategory had the most entries?
df[df["category"] == "consumables"]["subcategory"].value_counts().head(1)

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [68]:
#top_subcategory = df["category"].value_counts().idxmax()
#top_subcategory_by_category = df[df["category"] == top_subcategory]["subcategory"].value_counts().head(1)
#top_subcategory_by_category

In [71]:
# Which five clients had the most entries in the data?
top_5_clients = df["client_id"].value_counts()
most_entries = top_5_clients.head(5)
most_entries


client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [72]:
# Store the client ids of those top 5 clients in a list.
top_5_client_list = df.client_id.value_counts().head().index.to_list()
top_5_client_list

[33615, 66037, 46820, 38378, 24741]

In [75]:
# How many total units (the qty column) did the client with the most entries order order?
top_client = df["client_id"].value_counts().idxmax()
total_units = df[df["client_id"] == top_client]["qty"].sum()
total_units

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [27]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df["line_subtotal"] = df["qty"] * df["unit_price"]
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  line_subtotal  
0     1096.80     762.71         7.50  105            1      115164.00  
1       24.95      15.09         1.49   21            0         523.95  
2       13.52       7.86         1.68   39            6         527.28  
3       36.42      24.85         1.23   29            3        1056.18  
4      195.10     108.17        46.43   20            1        3902.00

In [28]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
df["total_weight"] = df["qty"] * df["unit_weight"]
df["shipping_price"] = df["total_weight"].apply(lambda x: x * 7 if x > 50 else x * 10)
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...     category         subcategory unit_price  \
0          17        2023  ...        decor            wall art    1096.80   
1          20        2023  ...  consumables                pens      24.95   
2           4        2023  ...     software  project management      13.52   
3          17        2023  ...  consumables                pens      36.42   
4          21        2023  ...  consumables                misc     195.10   

   unit_cost  unit_weight  qty  line_number  line_subtotal  total_weight  \
0     762.71         7.50  105            1      115164.00        787.50   
1      15.09         1.49   21            0         523.95         31.29   
2       7.86         1.68   39            6         527.28         65.52   
3      24.85         1.23   29            3        1056.18         35.67   
4     108.17        46.43   20            1        3902.00        928.60   

   shipping_price  
0         5512.50  
1          312.90  
2          458.64  
3          356.70  
4         6500.20  

[5 rows x 21 columns]

In [30]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df["line_price_before_tax"] = df["line_subtotal"] + df["shipping_price"]
df["tax_price"] = df["line_price_before_tax"] * 0.0925
df["line_price"] = df["line_price_before_tax"] + df["tax_price"]
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   
3          17        2023  ...     24.85        1.23   29            3   
4          21        2023  ...    108.17       46.43   20            1   

   line_subtotal  total_weight  shipping_price  line_price_before_tax  \
0      115164.00        787.50         5512.50              120676.50   
1         523.95         31.29          312.90                 836.85   
2         527.28         65.52          458.64                 985.92   
3        1056.18         35.67          356.70                1412.88   
4        3902.00        928.60         6500.20               10402.20   

      line_price     tax_price  
0  131839.076250  11162.576250  
1     914.258625     77.408625  
2    1077.117600     91.197600  
3    1543.571400    130.691400  
4   11364.403500    962.203500  

[5 rows x 24 columns]

In [32]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df["line_cost_without_shipping"] = df["unit_cost"] * df["qty"]
df["line_cost"] = df["line_cost_without_shipping"] + df["shipping_price"]
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...  qty line_number line_subtotal  total_weight  \
0          17        2023  ...  105           1     115164.00        787.50   
1          20        2023  ...   21           0        523.95         31.29   
2           4        2023  ...   39           6        527.28         65.52   
3          17        2023  ...   29           3       1056.18         35.67   
4          21        2023  ...   20           1       3902.00        928.60   

   shipping_price  line_price_before_tax     line_price     tax_price  \
0         5512.50              120676.50  131839.076250  11162.576250   
1          312.90                 836.85     914.258625     77.408625   
2          458.64                 985.92    1077.117600     91.197600   
3          356.70                1412.88    1543.571400    130.691400   
4         6500.20               10402.20   11364.403500    962.203500   

   line_cost_without_shipping  line_cost  
0                    80084.55   85597.05  
1                      316.89     629.79  
2                      306.54     765.18  
3                      720.65    1077.35  
4                     2163.40    8663.60  

[5 rows x 26 columns]

In [33]:
# Create a column for the profit of each line using line cost and line price
df["line_profit"] = df["line_price"] - df["line_cost"]
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... line_number line_subtotal total_weight  \
0          17        2023  ...           1     115164.00       787.50   
1          20        2023  ...           0        523.95        31.29   
2           4        2023  ...           6        527.28        65.52   
3          17        2023  ...           3       1056.18        35.67   
4          21        2023  ...           1       3902.00       928.60   

   shipping_price  line_price_before_tax     line_price     tax_price  \
0         5512.50              120676.50  131839.076250  11162.576250   
1          312.90                 836.85     914.258625     77.408625   
2          458.64                 985.92    1077.117600     91.197600   
3          356.70                1412.88    1543.571400    130.691400   
4         6500.20               10402.20   11364.403500    962.203500   

   line_cost_without_shipping  line_cost   line_profit  
0                    80084.55   85597.05  46242.026250  
1                      316.89     629.79    284.468625  
2                      306.54     765.18    311.937600  
3                      720.65    1077.35    466.221400  
4                     2163.40    8663.60   2700.803500  

[5 rows x 27 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [37]:
# Check your work using the totals above
order_id = [2742071, 2173913, 6128929]
filter_df = df[df["order_id"].isin(order_id)]
total_prices = filter_df.groupby("order_id")["line_price"].sum().reset_index()
total_prices

order_id     line_price
0   2173913  162388.719300
1   2742071  152811.897075
2   6128929  923441.240450

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [39]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
total_spend = df.groupby("client_id")["line_price"].sum().reset_index()
top_5_clients_by_profit = total_spend.sort_values(by = "line_price", ascending=False).head(5)
top_5_clients_by_profit

client_id    line_price
758      78965  6.559237e+08
840      85931  2.374170e+08
320      38473  1.835526e+08
850      86721  1.812663e+08
821      84344  1.066481e+08

In [57]:
top_5_clients_by_profit["line_price"].apply(lambda x: '%.2f' % x)

758    655923670.06
840    237417001.42
320    183552571.81
850    181266297.07
821    106648111.60
Name: line_price, dtype: object

In [40]:
filter_df_client = df[df["client_id"].isin(top_5_client_list)]
total_prices_client = filter_df_client.groupby("client_id")["line_price"].sum().reset_index()
total_prices_client

client_id    line_price
0      24741  8.226889e+07
1      33615  8.377309e+06
2      38378  1.290655e+07
3      46820  9.743794e+06
4      66037  1.025951e+07

In [49]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
totals = df.groupby('client_id').agg(
    total_units_purchased=('qty', 'sum'),
    total_costs=('line_cost', 'sum'),
    total_shipping_price=('shipping_price', 'sum'),
    total_revenue=('line_price', 'sum'),
    total_profit=('line_profit', 'sum')
).reset_index()
top_5_clients_summary = totals.sort_values(by='total_revenue', ascending=False).head(5)
top_5_clients_summary

client_id  total_units_purchased   total_costs  total_shipping_price  \
758      78965                4033302  4.085708e+08            5340756.88   
840      85931                 644423  1.281229e+08           10122549.19   
320      38473                 402770  1.402008e+08            4391003.73   
850      86721                1010426  1.382916e+08           64335727.72   
821      84344                 466968  7.893641e+07            4746583.13   

     total_revenue  total_profit  
758   6.559237e+08  2.473528e+08  
840   2.374170e+08  1.092941e+08  
320   1.835526e+08  4.335173e+07  
850   1.812663e+08  4.297470e+07  
821   1.066481e+08  2.771170e+07

In [55]:
# Format the data and rename the columns to names suitable for presentation.
totals = df.groupby('client_id').agg(
    Units=('qty', 'sum'),
    Shipping=('shipping_price', 'sum'),
    Revenue=('line_price', 'sum'),
    Cost=('line_cost', 'sum'),
    Profit=('line_profit', 'sum'),
).reset_index()
top_5_clients_summary = totals.sort_values(by='Revenue', ascending=False).head(5)
top_5_clients_summary

# Define the money columns. 
money_columns = ['total_shipping_price', 'total_revenue', 'total_profit']

# Define a function that converts a dollar amount to millions.
def currency_format_millions(amount):
    return '${:,.2f}M'.format(amount / 1000000)

# Apply the currency_format_millions function to only the money columns. 
top_5_clients_summary[money_columns] = top_5_clients_summary[money_columns].apply(currency_format_millions)

# Rename the columns to reflect the change in the money format. 
top_5_clients_summary.rename(columns={
    'total_units_purchased': 'Total Units Purchased',
    'total_shipping_price': 'Total Shipping Cost (Millions)',
    'total_revenue': 'Total Revenue (Millions)',
    'total_profit': 'Total Profit (Millions)'
}, inplace=True)

# Display the formatted DataFrame
print("Formatted Summary DataFrame:")
print(top_5_clients_summary)



KeyError: "None of [Index(['total_shipping_price', 'total_revenue', 'total_profit'], dtype='object')] are in the [columns]"

In [56]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
sorted_clients = top_5_clients_summary.sort_values(by = "Revenue", ascending=False)
sorted_clients.head()

client_id    Units     Shipping       Revenue          Cost        Profit
758      78965  4033302   5340756.88  6.559237e+08  4.085708e+08  2.473528e+08
840      85931   644423  10122549.19  2.374170e+08  1.281229e+08  1.092941e+08
320      38473   402770   4391003.73  1.835526e+08  1.402008e+08  4.335173e+07
850      86721  1010426  64335727.72  1.812663e+08  1.382916e+08  4.297470e+07
821      84344   466968   4746583.13  1.066481e+08  7.893641e+07  2.771170e+07